src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/HttpContentJsonExtensions.cs

```diff
@@ -0,0 +1,86 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Net.Mime;
+using System.Text;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http.Json
+{
+    public static class HttpContentJsonExtensions
+    {
+        public static Task<object?> ReadFromJsonAsync(this HttpContent content, Type type, JsonSerializerOptions? options = null, CancellationToken cancellationToken = default)
+        {
+            return ReadFromJsonAsyncCore(content, type, options, cancellationToken);
+        }
+
+        public static Task<T> ReadFromJsonAsync<T>(this HttpContent content, JsonSerializerOptions? options = null, CancellationToken cancellationToken = default)
+        {
+            return ReadFromJsonAsyncCore<T>(content, options, cancellationToken);
+        }
+
+        private static async Task<object?> ReadFromJsonAsyncCore(HttpContent content, Type type, JsonSerializerOptions? options, CancellationToken cancellationToken)
+        {
+            byte[] contentBytes = await GetUtf8JsonBytesFromContentAsync(content, cancellationToken).ConfigureAwait(false);
+            return JsonSerializer.Deserialize(contentBytes, type, options);
+        }
+
+        private static async Task<T> ReadFromJsonAsyncCore<T>(HttpContent content, JsonSerializerOptions? options, CancellationToken cancellationToken)
+        {
+            byte[] contentBytes = await GetUtf8JsonBytesFromContentAsync(content, cancellationToken).ConfigureAwait(false);
+            return JsonSerializer.Deserialize<T>(contentBytes, options);
+        }
+
+        private static async Task<byte[]> GetUtf8JsonBytesFromContentAsync(HttpContent content, CancellationToken cancellationToken)
+        {
+            string? mediaType = content.Headers.ContentType?.MediaType;
+
+            if (mediaType != JsonContent.JsonMediaType &&
+                mediaType != MediaTypeNames.Text.Plain)
+            {
```

Comment:
As per the [spec](https://github.com/dotnet/designs/blob/master/accepted/2020/json-http-extensions/json-http-extentions.md#whats-the-behavior-of-content-type), we allow `application/json` and `text/plain` only.

If Content-Type is null we assume it as "application/octet-stream" in accordance with section 7.2.1 of the HTTP spec.
This is how Formatting API works as well.

And at the same time, it is contrary to how `HttpContent.ReadAsStringAsync` works (it allows null content-type and tries to read the content using UTF-8 for that case).

IMO, null Content-Type should default to `application/json; charset=utf-8`.

Reply:
> And at the same time, it is contrary to how HttpContent.ReadAsStringAsync works (it allows null content-type and tries to read the content using UTF-8 for that case).

I know that JSON does have some rules about BOMs (I believe writing them is forbidden, reading them is left up to implementer?). We should make sure we're consistent with that.

Reply:
The web is a pretty wild place 👍. There are specs but there are also lots of implementations that don't conform in subtle ways, so this is good advice (and similar to what we do on the server side with JSON). 

---

I would expect this to also support suffix content types. `application/<something>+json` - https://tools.ietf.org/html/rfc6839

In use by places like github: https://developer.github.com/v3/media/ - it's a newer spec, but lots of more serious and large scale kinds of API design will prefer this over `application/json`. 

We might be able to get away with ignoring `text/json` which had some usage long ago, but isn't to spec.

Reply:
I'm a little bit surprised about `text/plain`. I'm not sure I've seen that in use for JSON.

Reply:
I can't remember why I put it in the spec; I think it was based on feedback of someone because I wouldn't have been smart enough to put either of these mime types in. If you believe `text/plain` is wrong let's remove it. Do we consider adding additional mime types in the future as a breaking change? IOW, would people commonly catch exceptions can do something different if a given mime type isn't accepted/produced by the server?

Reply:
+1 on the support for suffix content type. As of now, you are unable to read [problem details responses](https://tools.ietf.org/html/rfc7807) that MVC produces for error responses. It has a `application/problem+json` content type.

Reply:
What I get from this is that we need to white-list the following:
* `application/json`
* `application/<something>-json`
* ~`text/plain`~

@rynowak @pranavkm Is that correct?

Reply:
Sounds right to me.

Reply:
Yup, this looks fine. ASP.NET Core has two implementations for calculating media-type subsets:

* https://github.com/dotnet/aspnetcore/blob/master/src/Mvc/Mvc.Core/src/Formatters/MediaType.cs#L286-L291
* https://github.com/dotnet/aspnetcore/blob/master/src/Http/Headers/src/MediaTypeHeaderValue.cs#L385-L396

The first has fewer code dependencies and might be easier to easier to copy to S.T.Json if you'd like to use it as an implementation.